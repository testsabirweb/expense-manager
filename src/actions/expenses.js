import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({//for redux store
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {//for firebase
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',////setting default data
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
      .then(() => {
        dispatch(removeExpense({ id }))
      })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
      .then(() => {
        dispatch(editExpense(id, updates))
      })
  }
}

//SET_EXPENSE     FOR REDUX STORE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})
////START_SET_EXPENSES            ///FOR FETCHING DATA FROM FIREBASE
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = []
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          })
        })
        dispatch(setExpenses(expenses))
      })
  }
}