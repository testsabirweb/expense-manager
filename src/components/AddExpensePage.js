const React = require('react');
const ExpenseForm = require('./ExpenseForm')
const { connect } = require('react-redux')
const { addExpense } = require('./../actions/expenses')

const AddExpensePage = (props) => {
    return (
        <div>
            <h1>add expense</h1>
            <ExpenseForm
                onSubmit={(expense) => {
                    props.dispatch(addExpense(expense))
                    props.history.push('/')//for redirection to home page
                }}
            />
        </div>
    )
}

const ConnectedAddExpensePage = connect()(AddExpensePage)
module.exports = ConnectedAddExpensePage