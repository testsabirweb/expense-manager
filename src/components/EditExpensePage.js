const React = require('react');
const { connect } = require('react-redux')
const ExpenseForm = require('./ExpenseForm')
const { editExpense, startRemoveExpense } = require('./../actions/expenses')

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push('/')//for redirectonto home page
    }
    onClick = (e) => {
        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push('/')//for redirectonto home page
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onClick}
                > REMOVE
                </button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => {
            return expense.id === props.match.params.id
        })
    }
}
const mapDispatchToProps = (dispatch,props) => {
    return {
        editExpense:(id,expense)=>dispatch(editExpense(id,expense)),
        startRemoveExpense:(data)=>dispatch(startRemoveExpense(data))
    }
}

export default connect(mapStateToProps,mapDispatchToProps)(EditExpensePage)