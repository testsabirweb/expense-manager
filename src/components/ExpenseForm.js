const React = require('react')
const moment = require('moment')
const { SingleDatePicker } = require('react-dates')

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? props.expense.amount.toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => {
            return {
                description
            }
        })
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => {
            return {
                note
            }
        })
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => {
                return {
                    amount
                }
            })
        }
    }
    //https://github.com/airbnb/react-dates
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => {
                return {
                    createdAt
                }
            });
        }
    };
    //https://github.com/airbnb/react-dates
    onFocusChange = ({ focused }) => {
        this.setState(() => {
            return {
                calendarFocused: focused
            }
        });
    };
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => {
                return {
                    error: 'please provide description and amount'
                }
            })
        }
        else {
            this.setState(() => {
                return {
                    error: ''
                }
            })
            this.props.onSubmit({///this props can come from AddExpensePage or EditExpensePage
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),//valueOf()is a method from moment.js which returns the unix time stamp
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="add note to expense(optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button>add expense</button>
                </form>
            </div>
        )
    }
}

module.exports = ExpenseForm