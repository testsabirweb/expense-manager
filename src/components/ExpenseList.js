const React = require('react')
const { connect } = require('react-redux')
const ExpenseListItem = require('./ExpenseListItem')
const selectExpenses = require('./../selectors/expenses')


export const ExpenseList = (props) => {
    return (
        <div>
            {
                props.expenses.length === 0 ? (
                    <p>No expenses</p>) : (
                        props.expenses.map((expense) => {
                            return <ExpenseListItem key={expense.id} {...expense} />
                        }))
            }
        </div>
    )
}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
export default connect(mapStateToProps)(ExpenseList)

// const ConnectedExpenseList = connect((state) => {
//     return {
//         name: 'sabir'
//     }
// })(ExpenseList)

// module.exports = ConnectedExpenseList