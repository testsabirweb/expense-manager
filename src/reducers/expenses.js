
const expensesReducersDefault = []

const expensesReducer = (state = expensesReducersDefault, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ]
        case 'REMOVE_EXPENSE':
            const temp = state.filter(({ id }) => {
                return id !== action.id
            })
            return temp
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                }
                else {
                    return expense
                }
            })
        default:
            return state;
    }
}

module.exports=expensesReducer