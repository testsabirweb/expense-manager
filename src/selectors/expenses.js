const moment = require('moment')
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate, category }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
        const categoryMatch = (category === 'all' || category === expense.category) ? true : false
        return startDateMatch && endDateMatch && textMatch && categoryMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })
}

module.exports = getVisibleExpenses