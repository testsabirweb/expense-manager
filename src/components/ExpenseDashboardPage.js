const React = require('react');
import ExpenseList from './ExpenseList'
import ExpenseListFilters from'./ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => {
    return (
        <div>
            <h1>
                <ExpensesSummary />
                <ExpenseListFilters />
                <ExpenseList />
            </h1>
        </div>
    )
}

export default ExpenseDashboardPage
