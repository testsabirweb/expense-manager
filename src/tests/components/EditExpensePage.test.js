import React from 'react'
import {shallow} from 'enzyme'
import expenses from './../fixtures/expenses'
import {EditExpensePage }from './../../components/EditExpensePage'

let editExpenseSpy,removeExpenseSpy,wrapper,historySpy;

beforeEach(()=>{
    editExpenseSpy=jest.fn();
    removeExpenseSpy=jest.fn();
    historySpy={push:jest.fn()};
    wrapper=shallow(
        <EditExpensePage 
            editExpense={editExpenseSpy}
            removeExpense={removeExpenseSpy}
            history={historySpy}
            expense={expenses[2]}
        />
    )
})
test('should render ', () => {
    expect(wrapper).toMatchSnapshot()
})

