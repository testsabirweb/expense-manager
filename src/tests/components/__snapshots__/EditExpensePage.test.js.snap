// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render  1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 45,
        "createdAt": 345600000,
        "description": "coffee",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     REMOVE
  </button>
</div>
`;
